qutebrowser/misc/mhtml.py

```diff
@@ -0,0 +1,411 @@
+        if url in self.loaded_urls:
+            return
+        self.loaded_urls.add(url)
+
+        log.misc.debug("loading asset at %s", url)
+
+        download_manager = objreg.get("download-manager", scope="window",
+                                      window="current")
+        item = download_manager.get(url, fileobj=_NoCloseBytesIO(),
+                                    auto_remove=True)
+        self.pending_downloads.add((url, item))
+        item.finished.connect(
+            functools.partial(self.finished, url, item))
+        item.error.connect(
+            functools.partial(self.error, url, item))
+        item.cancelled.connect(
+            functools.partial(self.error, url, item))
+
+    def finished(self, url, item):
+        """Callback when a single asset is downloaded.
+
+        Args:
+            url: The original url of the asset as QUrl.
+            item: The DownloadItem given by the DownloadManager
+        """
+        self.pending_downloads.remove((url, item))
+        mime = item.raw_headers.get(b"Content-Type", b"")
+        mime = mime.decode("ascii", "ignore")
+
+        if mime.lower() == "text/css":
+            import_urls = _get_css_imports(item.fileobj.getvalue())
+            for import_url in import_urls:
+                import_url = import_url.decode("ascii")
+                absolute_url = url.resolved(QUrl(import_url))
+                self.fetch_url(absolute_url)
+
+        encode = E_QUOPRI if mime.startswith("text/") else E_BASE64
+        self.writer.add_file(url.toString(), item.fileobj.getvalue(), mime,
+                             encode)
+        item.fileobj.actual_close()
+        if self.pending_downloads:
+            return
+        self.finish_file()
+
+    def error(self, url, item, *_args):
+        """Callback when a download error occurred.
+
+        Args:
+            url: The orignal url of the asset as QUrl.
+            item: The DownloadItem given by the DownloadManager.
+        """
+        try:
+            self.pending_downloads.remove((url, item))
+        except KeyError:
+            # This might happen if .collect_zombies() calls .finished() and the
+            # error handler will be called after .collect_zombies
+            log.misc.debug("Oops! Download already gone: %s", item)
+            return
+        item.fileobj.actual_close()
+        self.writer.add_file(url.toString(), b"")
+        if self.pending_downloads:
+            return
+        self.finish_file()
+
+    def finish_file(self):
+        """Save the file to the filename given in __init__."""
+        if self._finished:
+            log.misc.debug("finish_file called twice, ignored!")
+            return
+        self._finished = True
+        log.misc.debug("All assets downloaded, ready to finish off!")
+        with open(self.dest, "wb") as file_output:
+            self.writer.write_to(file_output)
+        message.info("current", "Page saved as {}".format(self.dest), True)
```

Comment:
Sounds fair, I didn't know that! But in order to make the message appear, I have to click the titlebar again, even though the window is already focused (I never switched it), so it seemed like a strange behaviour.